src/miner.cpp

```diff
@@ -571,56 +554,93 @@ int BlockAssembler::UpdatePackagesForAdded(const CTxMemPool::setEntries& already
     return nDescendantsUpdated;
 }
 
-void BlockAssembler::RemoveTxIters(const std::vector<CTxMemPool::txiter> iters) {
+void BlockAssembler::RemoveEVMTransactions(const std::vector<CTxMemPool::txiter> iters) {
+
+    // Make sure we only remove EVM TXs which have no descendants or TX fees.
+    // Removing others TXs is more complicated and should be handled by RemoveFailedTransactions.
     for (const auto& iter : iters) {
         const auto &tx = iter->GetTx();
+        std::vector<unsigned char> metadata;
+        const auto txType = GuessCustomTxType(tx, metadata, false);
+        if (txType != CustomTxType::EvmTx) {
+            continue;
+        }
+
         for (size_t i = 0; i < pblock->vtx.size();  ++i) {
-            auto current = pblock->vtx[i];
-            if (current && current->GetHash() == tx.GetHash()) {
+            if (pblock->vtx[i] && pblock->vtx[i]->GetHash() == iter->GetTx().GetHash()) {
                 pblock->vtx.erase(pblock->vtx.begin() + i);
-                nBlockWeight -= iter->GetTxWeight();
-                nBlockSigOpsCost -= iter->GetSigOpCost();
```

Comment:
We are not removing sigopscost when removing failed transactions

Reply:
Indeed. #2220 is the attempt to fix all that correctly. This PR reverts and tries to be as close as possible to the original to get moving for now. 

(the sigops is OK since EVMTx do not hold sigscript)